### PR TITLE
event handling for 404 and 403 errorcodes

### LINKED ISSUE #10638: Handle 404 and 403 errors better in externalcluster controller
### What happened?

Handle 404 and 403 errors  better in externalcluster controller  to not reconcile external cluster controller when cluster  is not found or authorization failed.

### Expected behavior
Controller should instead emit events to inform user about the 404 and 403 errors.

### How is your environment configured?

- KKP version:
- Shared or separate master/seed clusters?:

### Provide your KKP manifest here (if applicable)

<details>

```yaml
# paste manifest here
```

</details>

### PR DESCRIPTION
**Pull request by imharshita**
Signed-off-by: Harshita sharma <harshita.sharma6174@gmail.com>

**What does this PR do / Why do we need it**:

**Does this PR close any issues?**:
Fixes #10638

**Special notes for your reviewer**:
Last commit reactors code to replace func argument from `ExternalClusterCloudSpec` => `ExternalClusterAKSCloudSpec` (for AKS), same for GKE and EKS as passing whole Spec is not required.

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

### CODE REVIEW

File: pkg/util/errors/errors.go [hunk 1]
```diff
@@ -21,6 +21,13 @@ import (
 	"net/http"
 )
 
+type StatusCode int
+
+const (
+	NotFoundStatusCode  StatusCode = 404
+	ForbiddenStatusCode StatusCode = 403
```
Comment (xrstf) [inline pkg/util/errors/errors.go]: You can just use the constants from `net/http`, no need to define your own set of HTTP status code constants.

File: pkg/provider/cloud/gke/provider.go [hunk 2]
```diff
@@ -479,3 +481,31 @@ func createClient(ctx context.Context, serviceAccount string, scope string) (*ht
...
+	if err == nil {
+		return statusCode
+	}
+
+	var apiErr *googleapi.Error
+	if errors.As(err, &apiErr) {
+		return apiErr.Code
+	}
+
+	return statusCode
```
Comment (xrstf) [inline pkg/provider/cloud/gke/provider.go]: Not sure why you created an unused variable here. This function can just be

```
func DecodeErrorCode(err error) int {
   var apiErr *googleapi.Error
   if errors.As(err, &apiErr) {
      return apiErr.Code
   }

   return 0
}
```

File: pkg/provider/cloud/gke/provider.go [hunk 3]
```diff
@@ -479,3 +481,31 @@ func createClient(ctx context.Context, serviceAccount string, scope string) (*ht
 
 	return client, projectID, nil
 }
+
+func DecodeError(err error) error {
+	if err == nil {
+		return nil
+	}
```
Comment (xrstf) [inline pkg/provider/cloud/gke/provider.go]: This nil check can be removed, I think. errors.As returns false if the error is nil: https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/errors/wrap.go;l=91

File: pkg/provider/cloud/eks/provider.go [hunk 4]
```diff
@@ -459,6 +460,23 @@ func DecodeAWSError(err error) error {
 	return err
 }
 
+func DecodeAWSErrorCode(err error) int {
+	var statusCode int
```
Comment (xrstf) [inline pkg/provider/cloud/eks/provider.go]: Same comment as for gke.DecodeErrorCode()

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve

**Pull request merged**